File: CrewRouterHelper/cr-report.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
cr-report —— CrewRouter 客户端事件统一上报器

所有 AI 客户端共用这一个程序，四种用法：

  1) hook 模式（Claude Code / Qwen Code / Codex 原生 hooks）：
       读 stdin 的 Claude 风格 hook JSON，映射后上报
       cr-report.py hook --harness claude_code
  2) emit 模式（Hermes / OpenClaw 等可直接执行命令的环境）：
       cr-report.py emit --harness hermes --event session_start --session <id>
  3) watch 模式（Grok 等无 hook 的客户端）：
       常驻 tail ~/.grok/sessions/**/updates.jsonl，增量解析为事件
       cr-report.py watch --harness grok
  4) test：发一条假事件验证链路
       cr-report.py test

配置文件：~/.config/cr-report.json（或环境变量 CR_REPORT_CONFIG 指定路径）
    { "url": "http://127.0.0.1:20003", "key": "cr-sk-..." }

铁律：任何失败都静默退出 0，绝不阻塞客户端工具执行。
"""

import argparse
import json
import os
import sys
import time
import urllib.request
import urllib.error

CONFIG_PATH = os.environ.get(
    "CR_REPORT_CONFIG", os.path.expanduser("~/.config/cr-report.json")
)
GROK_SESSIONS_DIR = os.path.expanduser("~/.grok/sessions")
STATE_PATH = os.path.expanduser("~/.cache/cr-report-grok-state.json")

# stdin JSON 的 hook_event_name -> 上报事件
EVENT_MAP = {
    "SessionStart": "session_start",
    "SessionEnd": "session_end",
    "PreToolUse": "tool_use",
    "PostToolUse": "tool_use",
}


def load_config():
    try:
        with open(CONFIG_PATH, "r", encoding="utf-8") as f:
            cfg = json.load(f)
        return str(cfg["url"]).rstrip("/"), str(cfg["key"])
    except Exception:
        return None, None


def post_event(cfg_url, cfg_key, payload):
    """静默上报；3 秒超时；任何异常吞掉。"""
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        cfg_url + "/api/client-events",
        data=data,
        headers={"Content-Type": "application/json",
                 "Authorization": "Bearer " + cfg_key},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=3):
            pass
    except Exception:
        pass


def cmd_hook(args):
    url, key = load_config()
    raw = ""
    try:
        raw = sys.stdin.read()
    except Exception:
        pass
    detail = {}
    try:
        detail = json.loads(raw) if raw.strip() else {}
    except Exception:
        detail = {"raw": raw[:512]}

    event = args.event
    if not event:
        hen = detail.get("hook_event_name") or detail.get("hookEventName") or ""
        event = EVENT_MAP.get(hen)
    if not event:
        # 无法判定事件类型时按工具调用处理（hook 场景下最常见）
        event = "tool_use" if (detail.get("tool_name") or detail.get("toolName")) else None
    if args.event == "session_start" or (not event and args.event is None):
        event = event or "session_start"

    payload = {
        "harness": args.harness,
        "event": event or "session_start",
        "session_id": detail.get("session_id") or detail.get("sessionId"),
        "tool_name": detail.get("tool_name") or detail.get("toolName"),
        "cwd": detail.get("cwd") or os.getcwd(),
        "ts": int(time.time()),
        "detail": {k: v for k, v in detail.items()
                   if k in ("hook_event_name", "source", "reason",
                            "tool_input", "message")},
    }
    if url and key:
        post_event(url, key, payload)
    return 0


def cmd_emit(args):
    url, key = load_config()
    payload = {
        "harness": args.harness,
        "event": args.event,
        "session_id": args.session,
        "tool_name": args.tool,
        "cwd": args.cwd or os.getcwd(),
        "ts": int(time.time()),
    }
    if url and key:
        post_event(url, key, payload)
    return 0


def _load_state():
    try:
        with open(STATE_PATH, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return {}


def _save_state(state):
    try:
        os.makedirs(os.path.dirname(STATE_PATH), exist_ok=True)
        with open(STATE_PATH, "w", encoding="utf-8") as f:
            json.dump(state, f)
    except Exception:
        pass


def scan_grok_files():
    """返回 {updates.jsonl 绝对路径: 文件大小}"""
    out = {}
    if not os.path.isdir(GROK_SESSIONS_DIR):
        return out
    for root, _dirs, files in os.walk(GROK_SESSIONS_DIR):
        if "updates.jsonl" in files:
            p = os.path.join(root, "updates.jsonl")
            try:
                out[p] = os.path.getsize(p)
            except OSError:
                pass
    return out


def parse_updates_line(line):
    """把一行 updates.jsonl 解析为 (event, tool_name) 或 None"""
    try:
        obj = json.loads(line)
    except Exception:
        return None
    t = obj.get("type") or obj.get("update_type") or ""
    if t == "agent_thought_chunk":
        return None  # 思维碎片不产生事件，太密
    if t == "tool_call":
        name = ((obj.get("payload") or {}).get("name")
                or obj.get("name") or "tool")
        return ("tool_use", str(name)[:128])
    if t == "agent_message_chunk":
        return None
    return None


def cmd_watch(args):
    url, key = load_config()
    if not url or not key:
        return 0
    state = _load_state()          # path -> 已读取的字节偏移
    known_sessions = set(state.keys())
    while True:
        try:
            files = scan_grok_files()
            # 新出现的会话目录 -> session_start
            for path in files:
                if path not in known_sessions:
                    sid = os.path.basename(os.path.dirname(path))
                    post_event(url, key, {
                        "harness": args.harness, "event": "session_start",
                        "session_id": sid[:128], "cwd": os.path.dirname(path),
                        "ts": int(time.time())})
                offset = state.get(path, 0)
                size = files[path]
                if size > offset:
                    try:
                        with open(path, "rb") as f:
                            f.seek(offset)
                            chunk = f.read(size - offset)
                        text = chunk.decode("utf-8", errors="replace")
                        lines = [ln for ln in text.splitlines() if ln.strip()]
                        # 半行保护：若文件未以换行结尾则回退偏移
                        if not text.endswith("\n"):
                            keep = len(lines[-1].encode("utf-8")) if lines else 0
                            state[path] = size - keep - 1 if size > keep else offset
                            lines = lines[:-1]
                        else:
                            state[path] = size
                        for ln in lines:
                            parsed = parse_updates_line(ln)
                            if parsed:
                                ev, tool = parsed
                                sid = os.path.basename(os.path.dirname(path))
                                post_event(url, key, {
                                    "harness": args.harness, "event": ev,
                                    "session_id": sid[:128], "tool_name": tool,
                                    "ts": int(time.time())})
                    except OSError:
                        pass
            known_sessions = set(files.keys())
            _save_state(state)
        except Exception:
            pass
        time.sleep(max(args.interval, 2))
    return 0


def cmd_test(args):
    url, key = load_config()
    if not url or not key:
        print(f"[cr-report] 配置缺失：{CONFIG_PATH}", file=sys.stderr)
        return 1
    payload = {
        "harness": args.harness or "hermes",
        "event": "session_start",
        "session_id": f"cr-report-test-{int(time.time())}",
        "cwd": os.getcwd(),
        "ts": int(time.time()),
        "detail": {"source": "cr-report test"},
    }
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        url + "/api/client-events", data=data,
        headers={"Content-Type": "application/json",
                 "Authorization": "Bearer " + key},
        method="POST")
    try:
        with urllib.request.urlopen(req, timeout=5) as resp:
            print(f"[cr-report] HTTP {resp.status} -> {resp.read().decode()}")
            return 0
    except Exception as e:
        print(f"[cr-report] 失败：{e}", file=sys.stderr)
        return 1


def main():
    ap = argparse.ArgumentParser(prog="cr-report")
    sub = ap.add_subparsers(dest="cmd", required=True)

    p_hook = sub.add_parser("hook", help="读 stdin hook JSON 并上报")
    p_hook.add_argument("--harness", required=True)
    p_hook.add_argument("--event", default=None,
                        help="强制指定事件；缺省从 stdin JSON 推断")
    p_hook.set_defaults(fn=cmd_hook)

    p_emit = sub.add_parser("emit", help="直接发一条事件")
    p_emit.add_argument("--harness", required=True)
    p_emit.add_argument("--event", required=True,
                        choices=["session_start", "session_end", "tool_use"])
    p_emit.add_argument("--session", default=None)
    p_emit.add_argument("--tool", default=None)
    p_emit.add_argument("--cwd", default=None)
    p_emit.set_defaults(fn=cmd_emit)

    p_watch = sub.add_parser("watch", help="常驻 tail Grok 会话目录")
    p_watch.add_argument("--harness", default="grok")
    p_watch.add_argument("--interval", type=int, default=5,
                         help="轮询秒数，默认 5")
    p_watch.set_defaults(fn=cmd_watch)

    p_test = sub.add_parser("test", help="发测试事件验证配置与链路")
    p_test.add_argument("--harness", default="hermes")
    p_test.set_defaults(fn=cmd_test)

    args = ap.parse_args()
    sys.exit(args.fn(args))


if __name__ == "__main__":
    try:
        main()
    except SystemExit:
        raise
    except Exception:
        # 兜底：绝不让上报器把宿主客户端搞挂
        sys.exit(0)

File: CrewRouterHelper/bin/cr-report.js
#!/usr/bin/env node
'use strict';
const fs = require('fs'); const path = require('path'); const readline = require('readline');
const api = require('../src');
function help() { console.log('用法: cr-report <hook|emit|status|login|logout|test|tui>'); }
function parse(args) { const out = {}; for (let i = 0; i < args.length; i++) { if (args[i].startsWith('--')) out[args[i].slice(2)] = args[i + 1]?.startsWith('--') ? true : args[++i]; else out._ = [...(out._ || []), args[i]]; } return out; }
async function hook(opts) { let detail = {}; try { const raw = fs.readFileSync(0, 'utf8'); detail = raw.trim() ? JSON.parse(raw) : {}; } catch {} const payload = api.normalizeHook(detail, opts.harness || 'grok', opts.event); if (payload.event) await api.report(payload); }
async function emit(opts) { const payload = { harness: opts.harness, event: opts.event, session_id: opts.session || null, tool_name: opts.tool || null, cwd: opts.cwd || process.cwd(), ts: Math.floor(Date.now() / 1000) }; await api.report(payload); }
async function login(opts) {
  const base = String(opts.url || process.env.CR_ROUTER_URL || '').replace(/\/$/, '');
  if (!base) throw new Error('需要 --url 或 CR_ROUTER_URL');
  const crypto = require('crypto'); const http = require('http'); const { execFile } = require('child_process');
  const verifier = crypto.randomBytes(48).toString('base64url');
  const challenge = crypto.createHash('sha256').update(verifier).digest('base64url');
  const state = crypto.randomBytes(16).toString('base64url');
  const server = http.createServer();
  await new Promise((resolve) => server.listen(0, '127.0.0.1', resolve));
  const port = server.address().port; const redirect = `http://127.0.0.1:${port}/callback`;
  const auth = new URL(`${base}/oauth/authorize`); for (const [key, value] of Object.entries({ client_id: 'crewrouter-helper', response_type: 'code', scope: 'events:report', redirect_uri: redirect, state, code_challenge: challenge, code_challenge_method: 'S256' })) auth.searchParams.set(key, value);
  console.log(`请在浏览器打开：\\n${auth}`);
  if (process.env.CR_REPORT_NO_BROWSER !== '1') { const opener = process.platform === 'darwin' ? 'open' : 'xdg-open'; execFile(opener, [String(auth)], () => {}); }
  const result = await new Promise((resolve) => { const timer = setTimeout(() => resolve({ error: 'timeout' }), 300000); server.on('request', (req, res) => { const u = new URL(req.url, 'http://127.0.0.1'); const value = { code: u.searchParams.get('code') || '', state: u.searchParams.get('state') || '', error: u.searchParams.get('error') || '' }; res.writeHead(200, { 'content-type': 'text/html; charset=utf-8' }); res.end(value.code && value.state === state ? '登录成功，请回到终端。' : '授权失败，请回到终端。'); clearTimeout(timer); resolve(value); }); });
  server.close(); if (result.error || !result.code || result.state !== state) throw new Error('授权失败或超时');
  const body = new URLSearchParams({ grant_type: 'authorization_code', code: result.code, client_id: 'crewrouter-helper', redirect_uri: redirect, code_verifier: verifier });
  const response = await new Promise((resolve, reject) => {
    const target = new URL(`${base}/oauth/token`); const transport = target.protocol === 'https:' ? require('https') : require('http');
    const req = transport.request(target, { method: 'POST', headers: { 'content-type': 'application/x-www-form-urlencoded', 'content-length': Buffer.byteLength(body) } }, (res) => { let text = ''; res.setEncoding('utf8'); res.on('data', (chunk) => { text += chunk; }); res.on('end', () => { try { resolve({ ok: res.statusCode >= 200 && res.statusCode < 300, json: JSON.parse(text) }); } catch { reject(new Error('服务端返回非法 JSON')); } }); });
    req.on('error', reject); req.end(body.toString());
  });
  if (!response.ok) throw new Error('换取 token 失败'); const token = response.json;
  if (!token.access_token || !token.refresh_token) throw new Error('服务端返回的凭证不完整');
  api.saveConfig({ url: base, client_id: 'crewrouter-helper', access_token: String(token.access_token), refresh_token: String(token.refresh_token), expires_at: Date.now() / 1000 + Number(token.expires_in || 86400), scope: String(token.scope || 'events:report') });
  console.log(`登录成功，凭证已写入 ${api.configPath()}（权限 600）`);
}
function logout() { const p = api.configPath(); try { fs.unlinkSync(p); console.log(`已删除 ${p}`); } catch (err) { if (err.code === 'ENOENT') console.log(`未找到本地凭证（${p}）`); else throw err; } }
async function test(opts) { const { url, token } = api.getCredential(); if (!url || !token) { console.error(`配置缺失：${api.configPath()}`); return 1; } const ok = await api.postEvent(url, token, { harness: opts.harness || 'grok', event: 'session_start', session_id: `cr-report-test-${Date.now()}`, cwd: process.cwd(), ts: Math.floor(Date.now() / 1000), detail: { source: 'cr-report test' } }, 5000); console.log(ok ? 'HTTP 上报成功' : 'HTTP 上报失败'); return ok ? 0 : 1; }
async function main() { const opts = parse(process.argv.slice(2)); const cmd = opts._?.[0]; if (!cmd || cmd === 'help' || opts.help) { help(); return 0; } if (cmd === 'hook') return hook(opts); if (cmd === 'emit') return emit(opts); if (cmd === 'status') { console.log(api.formatStatus(api.scanStatus(process.argv[1]))); return 0; } if (cmd === 'tui') { console.clear(); console.log(api.formatStatus(api.scanStatus(process.argv[1]))); console.log('\n按 Ctrl+C 退出；状态为只读扫描。'); readline.createInterface({ input: process.stdin, output: process.stdout }).on('close', () => {}); return 0; } if (cmd === 'login') return login(opts); if (cmd === 'logout') return logout(); if (cmd === 'test') return test(opts); if (cmd === 'hooks') { const action = opts._?.[1]; if (action === 'install') { console.log(`已安装 ${api.install(`${process.argv[1]} hook --harness grok`)}`); return 0; } if (action === 'uninstall') { console.log(api.uninstall() ? '已卸载 CrewRouter Hook' : '未找到 CrewRouter Hook'); return 0; } } help(); return 1; }
main().then((code) => { if (Number.isInteger(code)) process.exitCode = code; }).catch(() => { process.exitCode = 0; });
